### PR TITLE
Fix polygamma for negative non-integer inputs via reflection formula

### LINKED ISSUE #37635: Incorrect gradient of lax.digamma for negative half-integer inputs
## Description
`jax.lax.digamma` appears to produce an incorrect reverse-mode gradient for negative half-integer inputs.

The derivative of the digamma function is the trigamma function, so `jax.jacrev(lambda x: jax.lax.digamma(x))` should agree with `scipy.special.polygamma(1, x)` for finite non-pole inputs.

The tested values below are negative half-integers, not poles of the digamma function. SciPy returns finite positive trigamma values, but JAX returns small negative gradients.

## Minimal reproducible example:

```
import jax
import jax.numpy as jnp
from scipy.special import polygamma

fn = lambda x: jax.lax.digamma(x)

for v in [-16.5, -19.5, -22.5]:
    a = jnp.array([v])
    grad = float(jax.jacrev(fn)(a)[0, 0])
    expected = float(polygamma(1, v))
    print(f"digamma'({v})  jax={grad:+.4e}  scipy={expected:+.4f}")
```

Actual result:

```
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
digamma'(-16.5)  jax=-5.8807e-02  scipy=+9.8108
digamma'(-19.5)  jax=-4.9990e-02  scipy=+9.8196
digamma'(-22.5)  jax=-4.3471e-02  scipy=+9.8261
```

Expected result:

```
The gradient of jax.lax.digamma should match scipy.special.polygamma(1, x).

Expected values are approximately:
digamma'(-16.5) ~= +9.8108
digamma'(-19.5) ~= +9.8196
digamma'(-22.5) ~= +9.8261
```

### System info (python version, jaxlib version, accelerator, etc.)

```
WARNING:2026-05-07 00:33:38,915:jax._src.xla_bridge:794: An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.2.6
python: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0]
device info: cpu-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='MSI', release='5.15.167.4-microsoft-standard-WSL2', version='#1 SMP Tue Nov 5 00:21:55 UTC 2024', machine='x86_64')

nvidia-smi:
NVIDIA-SMI 545.23.06
Driver Version: 545.92
CUDA Version: 12.3
GPU: NVIDIA GeForce RTX 3080, 8192 MiB

Note: Reproduced on CPU fallback because CUDA-enabled jaxlib is not installed.
```

### PR DESCRIPTION
**Pull request by Sumu004**
`lax.polygamma(m, x)` and `digamma`'s gradient are badly wrong for negative non-integer `x` — XLA's `chlo.polygamma` uses an Euler-Maclaurin expansion that only converges for positive arguments, and diverges (sometimes catastrophically) for negative `x`.

For `x < 0`, applies the reflection formula obtained by differentiating `ψ(1-x) - ψ(x) = π·cot(πx)` `m` times, so the underlying `chlo.polygamma(m, 1-x)` call always sees a positive argument. Covers `m = 0, 1, 2` (digamma gradient, trigamma, tetragamma); higher orders are a natural follow-up.

Fixes #37635

### CODE REVIEW

**Comment by Sumu004** [comment]
Thanks for the review — both issues are addressed in the latest push:

**Issue 1 (m ≥ 3 fell back to g₂):** Added closed-form correction terms for m = 3 and m = 4, derived by repeated differentiation of g₀ = π·cot(πx):
- g₃ = −2π⁴(2 + cos(2πx)) / sin⁴(πx)
- g₄ = 8π⁵ cos(πx)(2 + cos²(πx)) / sin⁵(πx)

For m ≥ 5 the code now falls back to `chlo.polygamma(m, x)` directly (old, inaccurate but honest behavior) instead of silently applying g₂.

Verification:
```
polygamma(3, -0.5)  was: -1.41  now: 193.41  (scipy: 193.41) ✓
polygamma(4, -3.3)  was: wrong  now: 9738.73 (scipy: 9738.73) ✓
```

**Issue 2 (NaN at negative integer poles):** Negative integer x are detected via `int32` round-trip (avoids `int64` / `jax_enable_x64` dependency) and routed to `result_direct` (which returns ±∞ via `chlo.polygamma`). The `sin(πx)` denominator is also given a safe dummy value (0.5) for those inputs to suppress NaN in the inactive reflected branch.

```
polygamma(1, -1.0)  was: 6.58e32  now: +inf  ✓
polygamma(1, -2.0)  was: 1.65e32  now: +inf  ✓
```

All 17 passing tests still pass.

**Comment by jakevdp** [comment]
Thanks for working on this – it's nice to see the improved accuracy on the JAX side, but this is not fundamentally a JAX issue: the issue is in the XLA polygamma evaluation. Before we add this kind of complexity on the JAX side, we should explore whether the XLA-side fix is viable – I believe the relevant code is here: https://github.com/openxla/xla/blob/174c77e6a92f3c9839c112cb1f9cd56d887ec959/xla/hlo/builder/lib/math.cc#L1975

The real downside of the fix in this PR is that it replaces a single HLO operation (`chlo.polygamma`) with a long sequence of HLO operations: this will have the downside of vastly increasing compilation times for JAX programs that use `polygamma`, even in the non-problematic range. Given that, I don't think the JAX-side approach will be viable.

**Comment by Sumu004** [comment]
Thanks for the clear explanation, @jakevdp — that makes sense. Replacing a single `chlo.polygamma` with a sequence of trig HLO ops on every call is not the right trade-off, regardless of the `select` guard.

I'll move the fix upstream to XLA ([`xla/hlo/builder/lib/math.cc`](https://github.com/openxla/xla/blob/174c77e6a92f3c9839c112cb1f9cd56d887ec959/xla/hlo/builder/lib/math.cc#L1975)) where the Euler-Maclaurin expansion lives, so the reflection formula is applied inside the single HLO op and costs nothing for positive inputs. Will open a PR there and link back.

Happy to keep this JAX PR open as a documented stopgap in the meantime, or close it — whatever the team prefers.

**Comment by Sumu004** [comment]
XLA-side fix is now up: **openxla/xla#43545**

The patch adds the reflection formula directly inside `Polygamma`'s `doit` lambda in `xla/hlo/builder/lib/math.cc`, following the exact same technique `Digamma` already uses for its `x < 0.5` case (line 679). When the XLA fix lands and propagates into JAX's pinned XLA revision, `chlo.polygamma` will handle negative inputs correctly as a single HLO op — zero overhead for the common positive-input path.

Happy to close this JAX PR once the XLA fix merges; or keep it open as a stopgap in the meantime.
